File: frontend/types/automation.ts
/**
 * Automation Types
 * 
 * TypeScript interfaces for automation system
 */

export enum TriggerType {
  SCHEDULE = 'schedule',
  WEBHOOK = 'webhook',
  EVENT = 'event',
  MANUAL = 'manual'
}

export enum AutomationStatus {
  ACTIVE = 'active',
  PAUSED = 'paused',
  DRAFT = 'draft',
  ARCHIVED = 'archived'
}

export enum RunStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  SUCCESS = 'success',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  TIMEOUT = 'timeout'
}

export interface WorkflowStep {
  type: 'agent_task' | 'http_request' | 'condition' | 'delay' | 'set_variable'
  config: Record<string, any>
}

export interface TriggerConfig {
  event_type?: string
  conditions?: Array<{
    operator: string
    field: string
    value: any
  }>
  logic?: 'AND' | 'OR'
  webhook_secret?: string
  auth_type?: 'none' | 'signature' | 'api_key'
}

export interface Automation {
  id: string
  org_id: string
  name: string
  description?: string
  status: AutomationStatus
  trigger_type: TriggerType
  schedule?: string
  timezone: string
  workflow: {
    steps: WorkflowStep[]
  }
  trigger_config?: TriggerConfig
  agent_id?: string
  last_run_at?: string
  last_run_status?: RunStatus
  next_run_at?: string
  run_count: number
  success_count: number
  failure_count: number
  created_at: string
  updated_at: string
}

export interface AutomationRun {
  id: string
  automation_id: string
  status: RunStatus
  trigger_type: TriggerType
  trigger_data?: Record<string, any>
  started_at?: string
  completed_at?: string
  current_step: number
  total_steps: number
  context?: Record<string, any>
  result?: Record<string, any>
  error?: string
  logs: Array<{
    timestamp: string
    message: string
    level: string
  }>
  duration_seconds?: number
  created_at: string
}

export interface AutomationTemplate {
  id: string
  name: string
  description: string
  category: string
  icon: string
  trigger_type: TriggerType
  schedule?: string
  workflow: {
    steps: WorkflowStep[]
  }
  trigger_config?: TriggerConfig
  variables?: string[]
}

export interface CreateAutomationRequest {
  name: string
  description?: string
  workflow: {
    steps: WorkflowStep[]
  }
  trigger_type: TriggerType
  schedule?: string
  trigger_config?: TriggerConfig
  timezone?: string
  agent_id?: string
  enabled?: boolean
}

export interface UpdateAutomationRequest {
  name?: string
  description?: string
  workflow?: {
    steps: WorkflowStep[]
  }
  schedule?: string
  trigger_config?: TriggerConfig
  timezone?: string
  enabled?: boolean
  status?: AutomationStatus
}

File: backend/app/models/automation.py
"""
Automation Models

TASK-310: Create Automation model
TASK-311: Add workflow definition field (JSON)
TASK-312: Add schedule field (cron format)
TASK-313: Add trigger configuration
TASK-314: Create AutomationRun model
TASK-315: Add run status tracking
"""

from sqlalchemy import Column, String, Boolean, DateTime, ForeignKey, Text, Enum as SQLEnum, Integer
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
import uuid
from datetime import datetime
import enum

from app.db.session import Base


class TriggerType(str, enum.Enum):
    """Trigger types for automations"""
    SCHEDULE = "schedule"  # Cron-based schedule
    WEBHOOK = "webhook"  # External webhook
    EVENT = "event"  # Internal event (e.g., agent completion)
    MANUAL = "manual"  # User-triggered


class AutomationStatus(str, enum.Enum):
    """Automation status"""
    ACTIVE = "active"
    PAUSED = "paused"
    DRAFT = "draft"
    ARCHIVED = "archived"


class RunStatus(str, enum.Enum):
    """Automation run status"""
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    CANCELLED = "cancelled"
    TIMEOUT = "timeout"


class Automation(Base):
    """
    Automation model
    
    TASK-310: Automation model
    TASK-311: Workflow definition (JSON)
    TASK-312: Schedule field (cron)
    TASK-313: Trigger configuration
    """
    __tablename__ = "automations"
    
    # Primary key
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    # Organization relationship
    org_id = Column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    
    # Basic info
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    
    # Status
    status = Column(
        SQLEnum(AutomationStatus),
        nullable=False,
        default=AutomationStatus.DRAFT
    )
    
    # Workflow definition (TASK-311)
    # JSON structure: { "steps": [ { "type": "...", "config": {...} } ] }
    workflow = Column(JSONB, nullable=False, default=dict)
    
    # Trigger configuration (TASK-313)
    trigger_type = Column(
        SQLEnum(TriggerType),
        nullable=False,
        default=TriggerType.MANUAL
    )
    
    # Schedule (TASK-312) - cron format
    schedule = Column(String(255), nullable=True)  # e.g., "0 9 * * MON-FRI"
    
    # Trigger config (additional settings)
    trigger_config = Column(JSONB, nullable=False, default=dict)
    
    # Agent to use (optional)
    agent_id = Column(
        UUID(as_uuid=True),
        ForeignKey("agents.id", ondelete="SET NULL"),
        nullable=True
    )
    
    # Metadata
    created_by = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )
    
    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=datetime.utcnow
    )
    
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow
    )
    
    last_run_at = Column(DateTime(timezone=True), nullable=True)
    next_run_at = Column(DateTime(timezone=True), nullable=True)
    
    # Statistics
    total_runs = Column(Integer, nullable=False, default=0)
    success_runs = Column(Integer, nullable=False, default=0)
    failed_runs = Column(Integer, nullable=False, default=0)
    
    # Relationships
    organization = relationship("Organization", back_populates="automations")
    agent = relationship("Agent", back_populates="automations")
    runs = relationship("AutomationRun", back_populates="automation", cascade="all, delete-orphan")
    created_by_user = relationship("User")
    
    def __repr__(self):
        return f"<Automation {self.name} ({self.status})>"


class AutomationRun(Base):
    """
    Automation run model
    
    TASK-314: AutomationRun model
    TASK-315: Run status tracking
    """
    __tablename__ = "automation_runs"
    
    # Primary key
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    # Automation relationship
    automation_id = Column(
        UUID(as_uuid=True),
        ForeignKey("automations.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    
    # Run status (TASK-315)
    status = Column(
        SQLEnum(RunStatus),
        nullable=False,
        default=RunStatus.PENDING,
        index=True
    )
    
    # Trigger info
    trigger_type = Column(String(50), nullable=False)
    trigger_data = Column(JSONB, nullable=False, default=dict)
    
    # Execution details
    started_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    
    # Workflow state
    current_step = Column(Integer, nullable=True)
    total_steps = Column(Integer, nullable=True)
    
    # Execution context
    context = Column(JSONB, nullable=False, default=dict)
    
    # Results
    result = Column(JSONB, nullable=True)
    error = Column(Text, nullable=True)
    
    # Logs
    logs = Column(JSONB, nullable=False, default=list)
    
    # Metadata
    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=datetime.utcnow,
        index=True
    )
    
    # Celery task ID
    task_id = Column(String(255), nullable=True, index=True)
    
    # Relationships
    automation = relationship("Automation", back_populates="runs")
    
    def __repr__(self):
        return f"<AutomationRun {self.id} ({self.status})>"
    
    def add_log(self, level: str, message: str, data: dict = None):
        """Add log entry to run"""
        log_entry = {
            "timestamp": datetime.utcnow().isoformat(),
            "level": level,
            "message": message,
            "data": data or {}
        }
        
        if self.logs is None:
            self.logs = []
        
        self.logs.append(log_entry)
    
    @property
    def duration_seconds(self) -> float:
        """Calculate run duration in seconds"""
        if self.started_at and self.completed_at:
            return (self.completed_at - self.started_at).total_seconds()
        return 0.0
    
    @property
    def is_running(self) -> bool:
        """Check if run is currently running"""
        return self.status == RunStatus.RUNNING
    
    @property
    def is_complete(self) -> bool:
        """Check if run is complete (success or failure)"""
        return self.status in [RunStatus.SUCCESS, RunStatus.FAILED, RunStatus.CANCELLED, RunStatus.TIMEOUT]
